File: mobile/android/app/src/main/java/ai/treer/mobile/domain/Attention.kt
package ai.treer.mobile.domain

data class FleetLists(
    val attention: List<Agent>,
    val working: List<Agent>,
    val idle: List<Agent>,
    val homeAttention: List<Agent>,
    val inboxOverflow: Int,
    val onlineMachines: Int,
    val machineCount: Int,
    val workingCount: Int,
    val blockedCount: Int,
    val idleCount: Int,
)

fun agentNeedsAttention(agent: Agent, machine: Machine?): Boolean {
    if (agent.kind == "app") return false
    val status = agent.status.lowercase()
    if (status in ATTENTION_STATUSES) return true
    val machineOnline = machine?.isOnline == true
    return !machineOnline && status in NON_TERMINAL_STATUSES
}

fun agentSummary(agent: Agent, machine: Machine?): String {
    val machineName = machine?.displayName ?: agent.serverId
    return "${agent.name} is ${agent.status} on $machineName"
}

fun classifyFleet(snapshot: Snapshot, homeAttentionLimit: Int = 8): FleetLists {
    val agents = snapshot.fleetAgents
    val attention = agents.filter { agentNeedsAttention(it, snapshot.machine(it.serverId)) }
    val working = agents.filter { it.status.lowercase() in WORKING_STATUSES }
    val idle = agents.filter { agent ->
        val status = agent.status.lowercase()
        status == "idle" || status == "unknown"
    }
    val homeAttention = attention.take(homeAttentionLimit)
    val overflow = (attention.size - homeAttentionLimit).coerceAtLeast(0)
    val online = snapshot.servers.count { it.isOnline }
    return FleetLists(
        attention = attention,
        working = working,
        idle = idle,
        homeAttention = homeAttention,
        inboxOverflow = overflow,
        onlineMachines = online,
        machineCount = snapshot.servers.size,
        workingCount = working.size,
        blockedCount = agents.count { it.status.equals("blocked", true) },
        idleCount = idle.size,
    )
}

File: mobile/ios/Treer/Core/Attention.swift
import Foundation

enum Attention {
    static let homeLimit = 8

    static func needsAttention(agent: AgentInfo, machine: ServerInfo?) -> Bool {
        if agent.status == .blocked || agent.status == .failed {
            return true
        }
        if machine?.isOnline != true && agent.status.isNonTerminal {
            return true
        }
        return false
    }

    static func items(in snapshot: WorkspaceSnapshot) -> [AgentInfo] {
        snapshot.agents
            .filter { agent in
                needsAttention(agent: agent, machine: snapshot.machine(for: agent))
            }
            .sorted { lhs, rhs in
                if lhs.updatedAt != rhs.updatedAt {
                    return lhs.updatedAt > rhs.updatedAt
                }
                return lhs.name.localizedCaseInsensitiveCompare(rhs.name) == .orderedAscending
            }
    }

    static func working(in snapshot: WorkspaceSnapshot) -> [AgentInfo] {
        snapshot.agents
            .filter { agent in
                guard snapshot.machine(for: agent)?.isOnline == true else { return false }
                return agent.status == .starting || agent.status == .working
            }
            .sorted { $0.updatedAt > $1.updatedAt }
    }

    static func idle(in snapshot: WorkspaceSnapshot) -> [AgentInfo] {
        snapshot.agents
            .filter { agent in
                guard snapshot.machine(for: agent)?.isOnline == true else { return false }
                return agent.status == .idle || agent.status == .unknown
            }
            .sorted { lhs, rhs in
                let leftMachine = snapshot.machine(for: lhs)?.displayName ?? lhs.serverId
                let rightMachine = snapshot.machine(for: rhs)?.displayName ?? rhs.serverId
                if leftMachine != rightMachine {
                    return leftMachine.localizedCaseInsensitiveCompare(rightMachine) == .orderedAscending
                }
                return lhs.name.localizedCaseInsensitiveCompare(rhs.name) == .orderedAscending
            }
    }

    static func summary(for agent: AgentInfo, machine: ServerInfo?) -> String {
        let status = agent.displayStatus(machine: machine)
        let machineName = machine?.displayName ?? agent.serverId
        return "\(agent.name) is \(status) on \(machineName)"
    }

    static func homeAttention(in snapshot: WorkspaceSnapshot) -> (items: [AgentInfo], overflow: Int) {
        let all = items(in: snapshot)
        if all.count <= homeLimit {
            return (all, 0)
        }
        return (Array(all.prefix(homeLimit)), all.count - homeLimit)
    }
}
